### PR TITLE
Check web.xml version support against App Engine facet

### PR DESCRIPTION
**Pull request by briandealwis**
Relax the `web.xml` version validation from checking just `version="2.5"` to instead verify that it is compatible with the App Engine standard facet.  Required for Java 8 support (#1057).

### CODE REVIEW

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java [hunk 1]
```diff
@@ -66,7 +66,7 @@
     this.document = document;
     this.resource = resource;
     this.blacklist = new ArrayList<>();
-    validateJavaServlet();
+    validateJavaServletVersion();
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java]: `validateServletVersion`? That seems consistent with the methods below.

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java [hunk 2]
```diff
@@ -252,4 +252,12 @@ public boolean visit(IResource resource) throws CoreException {
     return webInfFolders;
   }
 
+  /**
+   * Check the the Dynamic Web Project facet version is compatible with the <tt>version</tt>.
+   */
+  public static boolean checkDeploymentDescriptorCompatibility(IProject project, String version) {
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java]: Probably does not belong to this PR.

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java [hunk 3]
```diff
@@ -252,4 +252,12 @@ public boolean visit(IResource resource) throws CoreException {
     return webInfFolders;
   }
 
+  /**
+   * Check the the Dynamic Web Project facet version is compatible with the <tt>version</tt>.
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java]: "Check that the ..."

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineStandardFacet.java [hunk 4]
```diff
@@ -64,6 +66,53 @@ public static boolean hasFacet(IFacetedProject project) {
...
+  }
+
+  /**
+   * Check that the given Servlet API version string (expected to be from the
+   * <tt>&lt;web-app version="xxx"&gt;</tt> attribute), is compatible with this project's App Engine
+   * settings.
+   * 
+   * @return {@code true} if supported or {@code false} otherwise
+   */
+  public static boolean checkServletApiSupport(IProject project, String servletApiVersion) {
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineStandardFacet.java]: I think it's straightforward to test this whole method, since you already have projects created from `TestProjectCreator`. Any reason for breaking down the method to test part of it?

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java (hunk 2 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java]: Oh geesh, I undid this change several times but it keeps coming back.

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineStandardFacet.java [hunk 5]
```diff
@@ -64,6 +66,53 @@ public static boolean hasFacet(IFacetedProject project) {
...
+  }
+
+  @VisibleForTesting
+  static boolean checkServletApiSupport(IProjectFacetVersion appEngineFacetVersion,
+      String servletApiVersion) {
+    // Assumes that App Engine Standard facet constraints are properly set up, so look up
+    // corresponding jst.web version and check for conflicts
+    IProjectFacetVersion dynamicWebFacetVersion = null;
+    try {
+      dynamicWebFacetVersion = WebFacetUtils.WEB_FACET.getVersion(servletApiVersion);
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineStandardFacet.java]: declare this here and move return statement inside try

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineStandardFacet.java [hunk 6]
```diff
@@ -64,6 +66,53 @@ public static boolean hasFacet(IFacetedProject project) {
...
+    } catch (IllegalArgumentException ex) {
+      // ignore: invalid version specified (e.g., "2.6")
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineStandardFacet.java]: return false

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java [hunk 7]
```diff
@@ -76,15 +76,16 @@
...
    */
-  private void validateJavaServlet() {
+  private void validateJavaServletVersion() {
     NodeList webAppList = document.getElementsByTagName("web-app");
     for (int i = 0; i < webAppList.getLength(); i++) {
       Element webApp = (Element) webAppList.item(i);
       String namespace = webApp.getNamespaceURI();
       String version = (String) webApp.getUserData("version");
       if ("http://xmlns.jcp.org/xml/ns/javaee".equals(namespace)
           || "http://java.sun.com/xml/ns/javaee".equals(namespace)) {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java]: namespace should match version

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineStandardFacet.java [hunk 8]
```diff
@@ -64,6 +66,53 @@ public static boolean hasFacet(IFacetedProject project) {
...
+   * Check that the given Servlet API version string (expected to be from the
+   * <tt>&lt;web-app version="xxx"&gt;</tt> attribute), is compatible with this project's App Engine
+   * settings.
+   * 
+   * @return {@code true} if supported or {@code false} otherwise
+   */
+  public static boolean checkServletApiSupport(IProject project, String servletApiVersion) {
+    IProjectFacetVersion appEngineFacetVersion = getProjectFacetVersion(project);
+    if (appEngineFacetVersion == null) {
+      return false;
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineStandardFacet.java]: If we can't find a facet, we don't want a marker. I.e. should this return true?

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java (hunk 7 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java]: #1834

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineStandardFacet.java (hunk 8 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineStandardFacet.java]: The validation already checks that the project has an App Engine Standard facet, so no marker will be created.  I'll turn this into a `Preconditions` check as it's ill-defined.
